### PR TITLE
Jsmc bpes

### PR DESCRIPTION
**Pull request by jsmcibm**
# Pull Request      

## What does this PR do?

adds BPES preprocessor - a preprocessor that (during training of MRC/tydi) uses the passage begin and end offsets as the truth when it encounters boolean questions

```Closes #(issue)```       

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            target=example['target']
+            passage_index=target['passage_indices'][0]
+            passage_start_offset=example['passage_candidates']['start_positions'][passage_index]
+            passage_end_offset=example['passage_candidates']['end_positions'][passage_index]
+
+            token_is_inside_passage = [passage_start_offset <= tok_start and
+                                        tok_end < passage_end_offset
+                                        for (tok_start, tok_end) in offsets]
+
+            # start/end in tokens is given by the first True and last True in after the query in token_is_inside_passage
```
Comment (seirasto) [inline primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py:45]: I'm not I understand this comment. Why can't you just take the `passage_start_offset` and `passage_end_offset`?

File: primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#                assert( passage_start_position <= start_position <= end_position <= passage_end_position)
+            features_dict['start_positions'] = passage_start_position
+            features_dict['end_positions'] = passage_end_position
+        # always return features_dict
+        return features_dict
+
+
+    def process_train(self, examples: Dataset) -> Tuple[Dataset, Dataset]:
+        """ 
+sample comparison of features and updated_features on Tydi train
```
Comment (seirasto) [inline primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py:67]: I don't know if this example is necessary. Perhaps an explanation would be more useful?

File: tests/primeqa/mrc/processors/preprocessors/test_tydiqa_bpes_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @pytest.fixture(scope='class')
+    def tydiqa_preprocessor(self, tokenizer):
+        return TyDiBoolQAPreprocessor(
+            tokenizer,
+            stride=128,
+            load_from_cache_file=False,
+        )
+
+    def test_train_preprocessing_runs_without_errors(self, train_examples, tydiqa_preprocessor):
+        train_examples, train_features = tydiqa_preprocessor.process_train(train_examples)
```
Comment (seirasto) [inline tests/primeqa/mrc/processors/preprocessors/test_tydiqa_bpes_preprocessor.py:37]: Should there be some checks to make sure nothing else other than for Y/N has been updated?

File: primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py (hunk 1 above)
Comment (mihaela-bornea) [reply primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py:45]: Are lines 31 and 32 needed?

File: tests/primeqa/mrc/processors/preprocessors/test_tydiqa_bpes_preprocessor.py (hunk 3 above)
Comment (mihaela-bornea) [reply tests/primeqa/mrc/processors/preprocessors/test_tydiqa_bpes_preprocessor.py:37]: I am OK with this test case

File: primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py (hunk 2 above)
Comment (jsmcibm) [reply primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py:67]: I strong feel that the example will be valuable if I need to debug this again

File: primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py (hunk 1 above)
Comment (seirasto) [reply primeqa/mrc/processors/preprocessors/tydiboolqa_bpes.py:45]: This looks good but I don't see the changes in the file?

**Review by mihaela-bornea — APPROVED** [review]
LGTM

**Review by seirasto — APPROVED** [review]
LGTM

**Pull request merged**